### PR TITLE
Mailet 101 (based on Mailet 100)

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesToMimeHeadersTest.java [hunk 1]
```diff
@@ -20,98 +20,65 @@
...
-        mci.setProperty("simplemapping", config2);
-        mailet.init(mci);
     }
-
-    private FakeMail setupMail(MimeMessage m) throws ParseException {
-        FakeMail mockedMail = MailUtil.createMockMail2Recipients(m);
+    
+
+    @Test
+    public void shouldPutAttributesIntoHeadersWhenMappingDefined() throws MessagingException {
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesToMimeHeadersTest.java]: What about testing when the attribute is already present?

Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesToMimeHeadersTest.java]: added

**Comment by chibenwa** [comment]
:+1:

**Comment by rouazana** [comment]
:+1:

**Comment by mbaechler** [comment]
updated with some other refactoring because some changes on FakeMail broke some other tests.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java [hunk 2]
```diff
@@ -20,72 +20,117 @@
 
 package org.apache.james.transport.mailets;
 
+import static org.mockito.Mockito.*;
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java]: Wildcard?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java [hunk 3]
```diff
@@ -20,72 +20,117 @@
...
+        Mail mail = MailUtil.createMockMail2Recipients(mimeMessage);
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader("header-name")).containsOnly("test-value", "first-value");
+    }
 
-        Assert.assertEquals(HEADER_VALUE, mockedMail.getMessage().getHeader(HEADER_NAME)[0]);
+    @Test
+    public void shouldNotThowOnMessagingException() throws MessagingException {
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java]: shouldNotThrow ? (missing "r")

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java [hunk 4]
```diff
@@ -20,56 +20,77 @@
...
-        assertNull(matchedRecipients);
-    }
-
-    protected Matcher createMatcher() {
-        return new SubjectStartsWith();
+    
+    @Test
+    public void shouldMatchWhenSubjectStartsWithConfiguredValue() throws MessagingException {
+        FakeMimeMessage mimeMessage = MailUtil.createMimeMessageWithSubject("testSubject");
+        FakeMail mail = new FakeMail();
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: Can't you use builder ?

```
FakeMail.builder()
    .mimeMessge(mimeMessage)
    .recipient(...)
    .build();
```

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java [hunk 5]
```diff
@@ -20,56 +20,77 @@
...
+        mail.setMessage(mimeMessage);
+        mail.setRecipients(Arrays.asList(new MailAddress("roger@nasa.org")));
+        
+        FakeMatcherConfig mailetConfig = new FakeMatcherConfig("SubjectIs=test", FakeMailContext.defaultContext());
+        
+        matcher.init(mailetConfig);
+
+        Collection<MailAddress> actual = matcher.match(mail);
+
+        assertThat(actual).isNotNull().hasSize(1).contains(new MailAddress("roger@nasa.org"));
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]:
```
assertThat(matcher.match(mail)).containsExactly(mailAddress)

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java [hunk 6]
```diff
@@ -20,56 +20,77 @@
...
+        FakeMimeMessage mimeMessage = MailUtil.createMimeMessageWithSubject(null);
+        FakeMail mail = new FakeMail();
+        mail.setMessage(mimeMessage);
+        mail.setRecipients(Arrays.asList(new MailAddress("roger@nasa.org")));
+        
+        FakeMatcherConfig mailetConfig = new FakeMatcherConfig("SubjectIs=test", FakeMailContext.defaultContext());
+        
+        matcher.init(mailetConfig);
+
+        Collection<MailAddress> actual = matcher.match(mail);
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: Idem

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java [hunk 7]
```diff
@@ -20,56 +20,77 @@
...
+        FakeMimeMessage mimeMessage = MailUtil.createMimeMessageWithSubject("foobar");
+        FakeMail mail = new FakeMail();
+        mail.setMessage(mimeMessage);
+        mail.setRecipients(Arrays.asList(new MailAddress("roger@nasa.org")));
+        
+        FakeMatcherConfig mailetConfig = new FakeMatcherConfig("SubjectIs=test", FakeMailContext.defaultContext());
+        
+        matcher.init(mailetConfig);
+
+        Collection<MailAddress> actual = matcher.match(mail);
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: Idem

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java (hunk 2 above)
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java]: eclipse neon changed it's default import refactoring, I need to fix that

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java (hunk 4 above)
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: the feature did not exist at the time I wrote that, but I will use it, ok

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java (hunk 3 above)
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMimeHeaderTest.java]: ok

**Comment by mbaechler** [comment]
I did a push force because I needed to rebase on master

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/AddSubjectPrefix.java [hunk 8]
```diff
@@ -57,6 +57,7 @@ public void init() throws MessagingException {
     public void service(Mail mail) throws MessagingException {
         MimeMessage m = mail.getMessage();
         String newSubject = prefixSubject(m);
+        m.setSubject(null);
```
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/AddSubjectPrefix.java]: You may extract those two lines in a method `replaceSubject`.

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java (hunk 5 above)
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: done

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java (hunk 6 above)
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: done

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java (hunk 7 above)
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SubjectStartsWithTest.java]: done

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/AddSubjectPrefix.java (hunk 8 above)
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/AddSubjectPrefix.java]: ok

**Comment by rouazana** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

**Pull request merged**